### PR TITLE
Feature/internal issue #601

### PR DESCRIPTION
**Pull request by Dronplane**
Updated NGRAM analyzer.
Added support for UTF-8 encoded strings.
Added support  for start/end ngram marking
Added tests.

### CODE REVIEW

File: core/analysis/ngram_token_stream.cpp [hunk 1]
```diff
@@ -229,11 +424,36 @@ bool ngram_token_stream::reset(const irs::string_ref& value) noexcept {
...
-
+  const size_t max_marker_size = std::max(options_.start_marker.size(), options_.end_marker.size());
+  if (max_marker_size > 0) {
+    // we have at least one marker. As we need to append marker to ngram and provide term
+    // value as continious buffer, we can`t return pointer to some byte inside input stream
+    // but rather we return pointer to buffer with copied values of ngram and marker
+    // For sake of performance we allocate requested memory right now
+    size_t buffer_size = options_.preserve_original ? data_.size() : std::min(data_.size(), options_.max_gram);
+    buffer_size += max_marker_size;
+    if (buffer_size > marked_term_buffer_.capacity()) {
```
Comment (gnusi) [inline core/analysis/ngram_token_stream.cpp]: ensured by `std::string::reserve(...)`

File: core/analysis/ngram_token_stream.cpp [hunk 2]
```diff
@@ -180,38 +277,136 @@ ngram_token_stream::ngram_token_stream(
...
-//FIXME UTF-8 support
+bool ngram_token_stream::next_symbol(const byte_type*& it) const noexcept {
+  if (it < data_end_) {
+    switch (options_.stream_bytes_type) {
+      case options_t::stream_bytes_t::BinaryStream:
+        ++it;
+        return true;
+      case options_t::stream_bytes_t::Ut8Stream:
+       {
+         uint8_t utf8_marker_byte = static_cast<uint8_t>(0xFF & *it);
```
Comment (gnusi) [inline core/analysis/ngram_token_stream.cpp]:
```suggestion
const byte_type utf8_marker_byte = *it;
```

File: core/analysis/ngram_token_stream.cpp [hunk 3]
```diff
@@ -180,38 +277,136 @@ ngram_token_stream::ngram_token_stream(
...
+         if (utf8_marker_byte < 0x80) {
+            ++it;
+          } else if ((utf8_marker_byte >> 5) == 0x6) {
```
Comment (gnusi) [inline core/analysis/ngram_token_stream.cpp]:
```suggestion
          } else if ((utf8_marker_byte >> 5) == 0x60) {
```

File: core/analysis/ngram_token_stream.cpp [hunk 4]
```diff
@@ -180,38 +277,136 @@ ngram_token_stream::ngram_token_stream(
...
+            it += 2;
+          } else if ((utf8_marker_byte >> 4) == 0xe) {
```
Comment (gnusi) [inline core/analysis/ngram_token_stream.cpp]:
```suggestion
          } else if ((utf8_marker_byte >> 4) == 0xE0) {
```

File: core/analysis/ngram_token_stream.cpp [hunk 5]
```diff
@@ -180,38 +277,136 @@ ngram_token_stream::ngram_token_stream(
...
+            it += 3;
+          } else if ((utf8_marker_byte >> 3) == 0x1e) {
```
Comment (gnusi) [inline core/analysis/ngram_token_stream.cpp]:
```suggestion
          } else if ((utf8_marker_byte >> 3) == 0x1E) {
```

File: core/analysis/ngram_token_stream.cpp [hunk 6]
```diff
@@ -164,14 +261,14 @@ NS_BEGIN(analysis)
...
-  REGISTER_ANALYZER_JSON(ngram_token_stream, make_json, 
+  REGISTER_ANALYZER_JSON(ngram_token_stream, make_json,
                          normalize_json_config); // match registration above
 }
 
 ngram_token_stream::ngram_token_stream(
     const options_t& options
 ) : analyzer(ngram_token_stream::type()),
-    options_(options) {
+    options_(options), start_marker_empty_(options.start_marker.empty()), end_marker_empty_(options.end_marker.empty()) {
```
Comment (gnusi) [inline core/analysis/ngram_token_stream.cpp]: add new line

File: core/analysis/ngram_token_stream.cpp [hunk 7]
```diff
@@ -180,38 +277,136 @@ ngram_token_stream::ngram_token_stream(
...
+            it += 4;
+          } else {
+            IRS_ASSERT(false);
+            IR_FRMT_ERROR("Invalid UTF-8 symbol increment");
+            return false;
+          }
+       }
+       return it <= data_end_;
+      default:
+        IRS_ASSERT(false);
```
Comment (gnusi) [inline core/analysis/ngram_token_stream.cpp]: Let's move this check to compile time, e.g.
```
enum class InputType { Binary, UTF8 };

template<InputType Type>
class ngram_token_stream {
  bool next_symbol(const byte_type&* it) {
    ...
    if /*contexpr*/ (Type == InputType::Binary) {
      ++it;
    } else if /*constexpr*/ (Type == InputType::UTF8) {
       it = utf8_next(it);
    }
    ...
  }
  ...
}
```

File: core/analysis/ngram_token_stream.cpp (hunk 1 above)
Comment (Dronplane) [reply core/analysis/ngram_token_stream.cpp]: Not until c++20

File: core/utils/utf8_utils.hpp [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+#ifndef IRESEARCH_UTF8_UTILS_H
+#define IRESEARCH_UTF8_UTILS_H
+
+#include "shared.hpp"
+#include "log.hpp"
+
+NS_ROOT
+NS_BEGIN(utf8_utils)
+
+inline  const byte_type* next(const byte_type* it, const byte_type* end) noexcept {
```
Comment (gnusi) [inline core/utils/utf8_utils.hpp]:
```suggestion
FORCE_INLINE const byte_type* next(const byte_type* it, const byte_type* end) noexcept {
```

File: core/analysis/analyzers.cpp [hunk 9]
```diff
@@ -208,7 +208,7 @@ NS_BEGIN(analysis)
 /*static*/ void analyzers::init() {
   #ifndef IRESEARCH_DLL
     irs::analysis::delimited_token_stream::init();
-    irs::analysis::ngram_token_stream::init();
+    irs::analysis::ngram_token_stream<irs::analysis::InputType::Binary>::init();
```
Comment (gnusi) [inline core/analysis/analyzers.cpp]: do we need?
```
irs::analysis::ngram_token_stream<irs::analysis::InputType::UTF8>::init();
```

**Comment by gnusi** [comment]
- drop `utf8` from iresearch
- check shared build

**Pull request merged**